File: android-vinetrack/app/src/main/java/com/rork/vinetrack/data/model/WorkTaskCostRollup.kt
package com.rork.vinetrack.data.model

import com.rork.vinetrack.data.material.WorkTaskMaterial
import com.rork.vinetrack.data.material.WorkTaskMaterialCosting
import com.rork.vinetrack.data.material.NullableMaterialNumericStringSerializer
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import java.math.BigDecimal
import java.math.RoundingMode

/** Stored authoritative cost allocation for one linked GPS trip slice. */
@Serializable
data class TripCostAllocation(
    val id: String,
    @SerialName("vineyard_id") val vineyardId: String,
    @SerialName("trip_id") val tripId: String,
    @SerialName("total_cost")
    @Serializable(with = NullableMaterialNumericStringSerializer::class)
    val totalCostRaw: String? = null,
    @SerialName("costing_status") val costingStatus: String? = null,
    @SerialName("deleted_at") val deletedAt: String? = null,
) {
    val totalCost: BigDecimal? get() = totalCostRaw?.toBigDecimalOrNull()
}

data class WorkTaskCostRollupResult(
    val labourCost: BigDecimal,
    val manualMachineCost: BigDecimal,
    val linkedTripCost: BigDecimal,
    val materialCost: BigDecimal,
    val totalCost: BigDecimal,
    val isComplete: Boolean,
) {
    fun costPerArea(areaHectares: Double, regionalAreaValue: Double): BigDecimal? {
        if (!isComplete || !areaHectares.isFinite() || areaHectares <= 0 || regionalAreaValue <= 0) return null
        return totalCost.divide(BigDecimal.valueOf(regionalAreaValue), 2, RoundingMode.HALF_UP)
    }
}

/** One shared, money-safe aggregation contract for every Work Task cost surface. */
object WorkTaskCostRollup {
    fun resolve(
        task: WorkTask,
        labourLines: List<WorkTaskLabourLine>,
        machineLines: List<WorkTaskMachineLine>,
        trips: List<Trip>,
        tripCostAllocations: List<TripCostAllocation>,
        materials: List<WorkTaskMaterial>,
        includeMaterials: Boolean,
    ): WorkTaskCostRollupResult {
        val scopedLabour = labourLines.filter { it.workTaskId == task.id && it.deletedAt == null }
        val effectiveLabour = PieceRateCosting.effectiveLabourCost(task, scopedLabour)
        val labourComplete = if (task.isPieceRate) {
            effectiveLabour != null
        } else {
            scopedLabour.all { it.hourlyRate != null || it.totalCost != null }
        }
        val labour = effectiveLabour?.let(BigDecimal::valueOf) ?: BigDecimal.ZERO

        val machine = machineLines
            .filter { it.workTaskId == task.id && it.deletedAt == null }
            .sumOf { BigDecimal.valueOf(it.resolvedCost) }
        val linkedTripIds = trips.filter { it.workTaskId == task.id }.mapTo(mutableSetOf()) { it.id }
        val linkedAllocations = tripCostAllocations.filter { it.deletedAt == null && it.tripId in linkedTripIds }
        val linked = linkedAllocations.mapNotNull { it.totalCost }.fold(BigDecimal.ZERO, BigDecimal::add)
        val material = if (includeMaterials) WorkTaskMaterialCosting.total(materials, task.id) else BigDecimal.ZERO
        val allocatedTripIds = linkedAllocations.mapTo(mutableSetOf()) { it.tripId }
        val linkedTripsComplete = linkedTripIds.all { it in allocatedTripIds } && linkedAllocations.all { it.totalCost != null }
        val complete = labourComplete && linkedTripsComplete

        return WorkTaskCostRollupResult(
            labourCost = labour.money(),
            manualMachineCost = machine.money(),
            linkedTripCost = linked.money(),
            materialCost = material.money(),
            totalCost = labour.add(machine).add(linked).add(material).money(),
            isComplete = complete,
        )
    }

    fun seasonTotal(
        tasks: List<WorkTask>,
        labourLines: List<WorkTaskLabourLine>,
        machineLines: List<WorkTaskMachineLine>,
        trips: List<Trip>,
        tripCostAllocations: List<TripCostAllocation>,
        materials: List<WorkTaskMaterial>,
        includeMaterials: Boolean,
    ): WorkTaskCostRollupResult {
        val results = tasks.map { resolve(it, labourLines, machineLines, trips, tripCostAllocations, materials, includeMaterials) }
        return WorkTaskCostRollupResult(
            labourCost = results.sumOf { it.labourCost }.money(),
            manualMachineCost = results.sumOf { it.manualMachineCost }.money(),
            linkedTripCost = results.sumOf { it.linkedTripCost }.money(),
            materialCost = results.sumOf { it.materialCost }.money(),
            totalCost = results.sumOf { it.totalCost }.money(),
            isComplete = results.all { it.isComplete },
        )
    }

    private fun BigDecimal.money(): BigDecimal = setScale(2, RoundingMode.HALF_UP)
}

File: ios/VineTrack/LegacyImported/Models/WorkTaskCostRollup.swift
import Foundation

/// One authoritative roll-up for every Work Task cost presentation.
/// Component engines retain their established semantics; this type only scopes
/// their saved inputs to one task and adds each authoritative source once.
nonisolated enum WorkTaskCostRollup {
    nonisolated struct Result: Equatable, Sendable {
        let labourCost: Decimal
        let manualMachineCost: Decimal
        let linkedTripCost: Decimal
        let materialCost: Decimal
        let totalCost: Decimal
        let isComplete: Bool

        func costPerHectare(areaHectares: Double) -> Decimal? {
            guard isComplete, areaHectares.isFinite, areaHectares > 0 else { return nil }
            return totalCost / WorkTaskCostRollup.decimal(areaHectares)
        }
    }

    static func resolve(
        task: WorkTask,
        labourLines: [WorkTaskLabourLine],
        machineLines: [WorkTaskMachineLine],
        trips: [Trip],
        tripCostAllocations: [TripCostAllocation],
        materials: [WorkTaskMaterial],
        includeMaterials: Bool
    ) -> Result {
        let taskLabourLines = labourLines.filter { $0.workTaskId == task.id }
        let labour: Double?
        let labourIsComplete: Bool
        if task.isPieceRate {
            labour = task.pieceRateCost
            labourIsComplete = labour != nil
        } else if taskLabourLines.isEmpty {
            labour = task.totalCost
            labourIsComplete = true
        } else {
            labour = WorkTaskLabourCosting.totalCost(taskLabourLines)
            labourIsComplete = taskLabourLines.allSatisfy { WorkTaskLabourCosting.lineCost($0) != nil }
        }

        let taskMachineLines = machineLines.filter { $0.workTaskId == task.id }
        // Preserve the established iOS contract: saved machine charge and saved
        // fuel are distinct values and both participate in the task total.
        let manualMachine = taskMachineLines.reduce(0.0) {
            $0 + ($1.totalMachineCost ?? 0) + ($1.fuelCost ?? 0)
        }

        let linkedTripIDs = Set(trips.filter { $0.workTaskId == task.id }.map(\.id))
        let linkedAllocations = tripCostAllocations.filter { linkedTripIDs.contains($0.tripId) }
        let linkedTrip = linkedAllocations.reduce(0.0) { $0 + ($1.totalCost ?? 0) }
        let allocatedTripIDs = Set(linkedAllocations.map(\.tripId))
        let tripIsComplete = linkedTripIDs.isSubset(of: allocatedTripIDs)
            && linkedAllocations.allSatisfy { $0.totalCost != nil }

        let material = includeMaterials
            ? WorkTaskMaterialCosting.total(materials, for: task.id)
            : 0
        let labourDecimal = decimal(labour ?? 0)
        let machineDecimal = decimal(manualMachine)
        let tripDecimal = decimal(linkedTrip)

        return combine(
            labourCost: labourDecimal,
            manualMachineCost: machineDecimal,
            linkedTripCost: tripDecimal,
            materialCost: material,
            isComplete: labourIsComplete && tripIsComplete
        )
    }

    static func combine(
        labourCost: Decimal,
        manualMachineCost: Decimal,
        linkedTripCost: Decimal,
        materialCost: Decimal,
        isComplete: Bool
    ) -> Result {
        Result(
            labourCost: labourCost,
            manualMachineCost: manualMachineCost,
            linkedTripCost: linkedTripCost,
            materialCost: materialCost,
            totalCost: labourCost + manualMachineCost + linkedTripCost + materialCost,
            isComplete: isComplete
        )
    }

    static func decimal(_ value: Double) -> Decimal {
        guard value.isFinite else { return 0 }
        return Decimal(string: String(value)) ?? 0
    }
}

@MainActor
extension WorkTask {
    /// Resolves this task from the store's durable child caches. Material cost is
    /// passed explicitly so callers cannot accidentally bypass the temporary gate.
    func costRollup(in store: MigratedDataStore, includeMaterials: Bool) -> WorkTaskCostRollup.Result {
        WorkTaskCostRollup.resolve(
            task: self,
            labourLines: store.workTaskLabourLines,
            machineLines: store.workTaskMachineLines,
            trips: store.trips,
            tripCostAllocations: store.tripCostAllocations,
            materials: store.workTaskMaterials,
            includeMaterials: includeMaterials
        )
    }
}
